node/neighbor.go

```diff
@@ -424,3 +425,85 @@ func (localNode *LocalNode) maybeAddRemoteNode(remoteNode *nnetnode.RemoteNode)
 
 	return nil
 }
+
+func (localNode *LocalNode) VerifySigChain(sc *pb.SigChain, height uint32) error {
+	c, ok := localNode.nnet.Network.(*chord.Chord)
+	if !ok {
+		log.Fatal("Overlay is not chord")
+	}
+
+	if config.SigChainVerifySkipNode.GetValueAtHeight(height) {
+		// only needs to verify node to node hop, and no need to check last node to
+		// node hop because it could be successor
+		for i := 1; i < sc.Length()-3; i++ {
+			dist := chord.Distance(sc.Elems[i].Id, sc.Elems[i+1].Id, config.NodeIDBytes*8)
+			fingerIdx := dist.BitLen() - 1
+			fingerStartID := chord.PowerOffset(sc.Elems[i].Id, uint32(fingerIdx), config.NodeIDBytes*8)
+			if !chord.BetweenLeftIncl(fingerStartID, sc.Elems[i+1].Id, localNode.Id) {
+				continue
+			}
+			skipped := 1
+			for _, succ := range c.Successors() {
+				if skipped >= por.NextHopChoice {
+					break
+				}
+				if chord.BetweenLeftIncl(fingerStartID, sc.Elems[i+1].Id, succ.Id) {
+					skipped++
+				}
+			}
+			for _, pred := range c.Predecessors() {
+				if skipped >= por.NextHopChoice {
+					break
+				}
+				if chord.BetweenLeftIncl(fingerStartID, sc.Elems[i+1].Id, pred.Id) {
+					skipped++
+				}
+			}
+			if skipped >= por.NextHopChoice {
+				return fmt.Errorf("skipped nodes")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (localNode *LocalNode) VerifySigChainObjection(sc *pb.SigChain, submitterID []byte, height uint32) error {
+	if !config.SigChainObjection.GetValueAtHeight(height) {
+		return fmt.Errorf("sigchain objection is not enabled")
+	}
+
+	if config.SigChainVerifySkipNode.GetValueAtHeight(height) {
+		// only needs to verify node to node hop, and no need to check last node to
+		// node hop because it could be successor
+		for i := 1; i < sc.Length()-3; i++ {
+			dist := chord.Distance(sc.Elems[i].Id, sc.DestId, config.NodeIDBytes*8)
+			fingerIdx := dist.BitLen() - 1
+			fingerStartID := chord.PowerOffset(sc.Elems[i].Id, uint32(fingerIdx), config.NodeIDBytes*8)
+			if !chord.BetweenLeftIncl(fingerStartID, sc.Elems[i+1].Id, submitterID) {
+				return fmt.Errorf("submitter is not skipped")
+			}
```

Comment:
submitterID should be one of sc.Elems[].id
Submitter who from out of relay participator is suspect.